Wikipedia/Code/WMFAppViewController.m

```diff
@@ -334,22 +355,49 @@ - (void)launchAppInWindow:(UIWindow *)window waitToResumeApp:(BOOL)waitToResumeA
 
     [self showSplashView];
 
-#if TEST_SHARED_CONTAINER_MIGRATION
-    NSFileManager *fm = [NSFileManager defaultManager];
-    [fm removeItemAtPath:[YapDatabase wmf_databasePath] error:nil];
-    [fm removeItemAtPath:[MWKDataStore mainDataStorePath] error:nil];
-    [fm removeItemAtPath:[SDImageCache wmf_imageCacheDirectory] error:nil];
-    [[NSUserDefaults wmf_userDefaults] wmf_setDidMigrateToSharedContainer:NO];
-#endif
+    [self migrateIfNecessary];
+}
+
+- (void)migrateIfNecessary {
+    if (self.isMigrationComplete || self.isMigrationActive) {
+        return;
+    }
+
+    __block BOOL migrationsAllowed = YES;
+    [self startMigrationBackgroundTask:^{
+        migrationsAllowed = NO;
+    }];
+    dispatch_block_t bail = ^{
+        [self endMigrationBackgroundTask];
+        self.migrationActive = NO;
+    };
     [self migrateToSharedContainerIfNecessaryWithCompletion:^{
```

Comment:
Ignore for now that this horrible chain of completions should really all be `NSOperations` in a serial queue.